### ISSUE TITLE
Etterfilter inject causes a loss of the modified packet (MQTT)

### ISSUE THREAD

**Issue by JaumeGarciaU** [Ettercap/ettercap#1139]
Hi eveyone!
I'm trying to use etterfilter to get and modifiy a MQTT packet.
I'm trying to simulate a connection between a sensor node and a central node.
The original connection is implemented with the following command

$mosquitto_pub -d -q 1 -h "192.168.88.247" -t "v1/devices/me/telemetry" -u "rWrmc7NwVD0GGnrqlifZ" -m "{"temperature":20, "humidity":20}"

If i execute the command without any filter applied the connection is succesful
![image](https://user-images.githubusercontent.com/37404255/131513968-ea362b12-005e-4b0d-8375-a56f08577008.png)

The problem i'm facing is that i would like to create a filter that modifies this temperature and humidity value. I know that i could use the replace( ) function, but later on i would like to be able to modify the value without knowing its original value. That's why i tought about the drop( ) and inject( ) sequence.

That's why i tried to implement the following filter
________
if (ip.proto == TCP && tcp.dst == 1883 && ip.dst == '192.168.88.247' &&
search(DATA.data, "{temperature:")) {
 
 #First we save the original packet for later checks and we drop it
 log(DECODED.data,"startData.log"); 
 drop();
 
 #We inject the desired value
 inject("injectedData.log");

 #Save the modified packet for later checks
 log(DATA.data,"endData.log"); 
 msg("payload replaced\n");
}
______

where the injectedData.log is the following file
28v1/devices/me/telemetry{temperature:15, humidity:55}

My idea is that the central node recieves the temperature:15, humidity:55 values insted of the temperature":20, "humidity":20
The data is recieved by the central node and the value is saved, but the connection steps end up failing.

Steps to reproduce
1. Compile the filter -> # etterfilter owned.filter -o owned.ef
2. Execute ettercap with the filter -> sudo ettercap -T -q -i eth0 -F owned.ef -M ARP /192.168.88.246// /192.168.88.247//
(where ip .246 is the sensor node and ip .247 the central node)
3. Send the request from the sensor node
4.  
![image](https://user-images.githubusercontent.com/37404255/131516677-7cf58422-7f62-4207-8649-21195f977f2d.png)

As i mentioned, the central host saves the modified data (Well, my idea was that only the modified packet arrives to the central host, instead both the original and modified packet arrives and the central host only saves the data of the modified one, but i leave that for later)

I've checked that if i use the file "startData.log" in the inject( ) command, the connection is succesful, so i think that maybe when i use the drop( ) and inject( ), i'm loosing some information of the session

Don't know if anyone could help me, any ideas will be highly appreciated (:

**Comment by koeppea** [comment]
Hi, can you please re-run ettercap with the `-w ettercap-packets.pcap` argument?
I'd like to understand what happens on the wire, at least from Ettercap's perspective.

**Comment by JaumeGarciaU** [comment]
Hi @koeppea !
This is the result i got
[ettercap-packets.pcap.gz](https://github.com/Ettercap/ettercap/files/7086958/ettercap-packets.pcap.gz)

**Comment by LocutusOfBorg** [comment]
packet 195 might be lacking a 0x0a at the end.
This causes a retrasmission of the packet in packet 197 probably with the original content?

I don't have any mqtt dump locally, so my analysis might be wrong

**Comment by JaumeGarciaU** [comment]
> packet 195 might be lacking a 0x0a at the end.
> This causes a retrasmission of the packet in packet 197 probably with the original content?
> 
> I don't have any mqtt dump locally, so my analysis might be wrong

But that's the result i have when i try to send the request without executing ettercap.
In my first message, the first screenshot results in the following "publish message" packet (equivalent to the 197 packet you mentioned)
![image](https://user-images.githubusercontent.com/37404255/131650291-2fb79d06-1557-49c3-9c11-109c4354def1.png)
Maybe is the 0x0a info that causes the loss of the packet that replies to 197?

**Comment by LocutusOfBorg** [comment]
yes, even wireshark can't detect the packet format, because the len is 1 byte more than the declared one. That 0a at the end is probably making the package look "wrong" for the analyzer too.
